### ISSUE TITLE
[PyTorch] VideoLoader gets stuck at pytorch/nvvl/loader.py:106 

### ISSUE THREAD

**Issue by Ivorra** [NVIDIA/nvvl#29]
Hi,

When loading certain sets of videos (order does not matter) into a PyTorch VideoLoader, and iterate through it, the VideoLoader silently stops working after some iterations, getting stuck.

Specifically I have seen that it stops [here](https://github.com/NVIDIA/nvvl/blob/master/pytorch/nvvl/loader.py#L106), line 106 in loader.py:

 &nbsp;&nbsp;&nbsp;&nbsp;`self.dataset._finish_receive()`

Regarding the parameters of both the VideoDataset and VideoLoader, this problem seems to be independent of the `sequence_length` or the `batch_size`, respectively. For the ProcessDesc object, I use `type="float"` and `color_space="RGB"`.

The videos (AVI files of the UCF-101 dataset) have been converted to MP4 with the provided command. In this case, the keyframe frequency is one:

 &nbsp;&nbsp;&nbsp;&nbsp;`ffmpeg -i v.avi -map v:0 -c:v libx264 -crf 18 -pix_fmt yuv420p -g 1 -profile:v high v.mp4`  

I have attached into a [Gist](https://gist.github.com/Ivorra/86a406841f8d99b339952352711f0635) the Python script I am using as well as a small dataset of four converted videos that arise the error.

Videos also [here](https://www.dropbox.com/s/7exizs8ry5i65xh/source_videos.zip?dl=0).

Thanks!

**Comment by pikerbright** [comment]
You can open debug log and see the frame num poped by decoder. This stuck issue is alway caused because it can't get the right frame num you needed(eg, you request the frame out of range of the video).

**Comment by pikerbright** [comment]
@Ivorra before you iterate through the loader, you need to get the total frame count for each video which is calculated as FPS * duration. But the FPS and duration may be not precise. So there is still a chance to get out of range.
I used to get a video only has 298 frame, but its fps is 30, duration is 10s.

**Comment by Ivorra** [comment]
In this case I am not manually iterating the loader, I am using:

```
for video in loader:
    do something
```


So no need for counting the number of frames.

By the way, the number of frames can also be counted with NVVL, by using:
```
lib_loader = nvvl.lib.nvvl_create_video_loader_with_log(
    gpu_id, nvvl.lib.LogLevel_Error)  # Or another LogLevel

num_frames = nvvl.lib.nvvl_frame_count(lib_loader, str.encode(path_to_video))
```

You have to take into account that this takes some time if you count many videos

**Comment by pikerbright** [comment]
Indeed, I actually mean that use nvvl_frame_count to get the frame num of video.
Did you use VideoDataset in dataset.py? you can see the code like:
`        for f in filenames:
            count = lib.nvvl_frame_count(self.loader, str.encode(f));
            count -= self.sequence_length + 1
`

**Comment by Ivorra** [comment]
Yes, I use dataset.py. Specifically, the code that I am used can be seen in this [Gist](https://gist.github.com/Ivorra/86a406841f8d99b339952352711f0635). In principle, I do not need to count the number of frames for iterating the VideoLoader. 

Are you suggesting that there is an alternative?

**Comment by jaredcasper** [comment]
Sorry for the delay.  It seems this is a case where the assumption that the code makes that the number of frames is equal to the frame rate times the duration is incorrect.  I'll have to do some research to see if there is a better way to determine the number of frames in the file (short of scanning the entire file). The calculation is done on line 271 of VideoLoader.cpp.   Short of that it should notice when the requested frame is past the end of the video and handle it rather than get stuck.  I'll try to get to this next week.

**Comment by Ivorra** [comment]
No problem, thanks for looking at it.

By the way, I tried to re-apply the same command to the videos it seemed the loader got stuck at (the ones in MP4 format).  

After repeating this a few times, in each ocassion it was a different video (the loader got stuck at a different video), the loader stopped getting stuck. 

Does this mean that FFmpeg is varying the number of frames? 

Thanks!

**Comment by mcever** [comment]
I'm having a similar problem where, with certain combinations of videos, the code will hang at

`self.dataset._finish_receive()`

when iterating using

`for data in dataloader:`

In my case, I'm testing with 5 videos. If I create a nvvl.VideoDataset using paths to vid0.mp4 and vid3.mp4, I am able to complete my feature extraction loop on each sample. Similarly, the loop terminates successfully using combinations of vid[1,2,4].mp4. However, I'm finding if I try to create a nvvl.VideoDataset using a combination including videos from the two sets, the loop will hang when it begins to process a file from another set.

This makes it seem like it could be a problem besides the frame rate calculation mentioned here, but it also seems similar enough that this could be the issue, especially considering that Ivorra is able to iterate over a nvvl.VideoDataset created with any given single video as well. Hopefully this helps.
